### ISSUE TITLE
`nextstrain update singularity` throws error about OCI indexes with Singularity 2.6 and 3.5.3

### ISSUE THREAD

**Issue by tsibley** [nextstrain/cli#267]
```
$ nextstrain update singularity
Checking for newer versions of Nextstrain CLI…

nextstrain-cli is up to date!

Updating singularity runtime…
Updating Singularity image from docker://nextstrain/base:build-20230119T003940Z to docker://nextstrain/base:build-20230314T190832Z…

Docker image path: index.docker.io/nextstrain/base:build-20230314T190832Z
ERROR MANIFEST_UNKNOWN: OCI index found, but accept header does not support OCI indexes
Cleaning up...

Updating failed!
```

Did we switch from a single platform image to a multi-platform image (i.e. image manifest) between these two versions? Is that what's causing Singularity grief here?

**Comment by tsibley** [comment]
@victorlin You may be interested in looking into this…

**Comment by victorlin** [comment]
Hmm, [build-20230119T003940Z](https://hub.docker.com/r/nextstrain/base/tags?page=1&name=build-20230119T003940Z) is already multi-platform (for all images created following https://github.com/nextstrain/docker-base/commit/960fb5bb8a7abdfbaaf75579019101d0b3067940) so that alone can't be the reason. But yes, manifest / OCI index is used for the multi-platform image.

Can you tell where this error is coming from or where the "accept header" is defined?

**Comment by tsibley** [comment]
Ah, ok, well there goes that theory, at least barring ways it might have gone unnoticed previously.

I'm sure I _can_ tell where it's coming from, but I haven't looked yet. :see_no_evil:

**Comment by tsibley** [comment]
Singularity (presumably) is making the following request:

```
GET /v2/nextstrain/base/manifests/build-20230321T203820Z
Host: index.docker.io
Accept: application/vnd.docker.distribution.manifest.v2+json,application/vnd.docker.distribution.manifest.list.v2+json,application/json
```

and Docker's registry responds with a 404 with following body:

```json
{"errors":[{"code":"MANIFEST_UNKNOWN","message":"OCI index found, but accept header does not support OCI indexes"}]}
```

Singularity (presumably) needs to be including the [OCI image index](https://github.com/opencontainers/image-spec/blob/main/image-index.md) media type in its request:

```
application/vnd.oci.image.index.v1+json
```

and be prepared to handle it in the response. An OCI image index serves the same purpose as Docker's manifest lists (`application/vnd.docker.distribution.manifest.list.v2+json`).

**Comment by tsibley** [comment]
I've been using Singularity 2.6.1, intentionally not the latest 3.x, in order to development against a still very prevalent older release series.

Upgrading to 3.11.1 instead makes this error go away, presumably because Singularity added OCI support in 3.x.

I guess our multi-platform image broke compat with Singularity 2.6? I guess that's ok given the other benefits, but good to know!

**Comment by tsibley** [comment]
> Hmm, [build-20230119T003940Z](https://hub.docker.com/r/nextstrain/base/tags?page=1&name=build-20230119T003940Z) is already multi-platform (for all images created following [nextstrain/docker-base@960fb5b](https://github.com/nextstrain/docker-base/commit/960fb5bb8a7abdfbaaf75579019101d0b3067940)) so that alone can't be the reason.

Hmm. Both are indeed multi-platform (amd64 and arm64), but the registry returns `application/vnd.docker.distribution.manifest.list.v2+json` for build-20230119T003940Z and `application/vnd.oci.image.index.v1+json` for `build-20230321T203820Z`. Maybe our method of producing the multi-platform images changed?

**Comment by tsibley** [comment]
A binary search between those two images reveals that the media type switched to OCI in `build-20230120T002454Z` (the very next image after the first).

**Comment by tsibley** [comment]
build-20230119T003940Z was built in https://github.com/nextstrain/docker-base/actions/runs/3954209788/jobs/6771310727.

build-20230120T002454Z was built in https://github.com/nextstrain/docker-base/actions/runs/3963573743/jobs/6791555853.

Both from the same source: d59452c4800c09e7f418c80d70379523a7cf84ef

The GitHub Actions runner image did change between the two:

    20230109.1
    20230118.2

**Comment by tsibley** [comment]
buildx changed 0.9.1 → 0.10.0 between those two runner images. [0.10.0 enabled a "provenance attestation" feature by default](https://github.com/docker/buildx/releases/tag/v0.10.0), which requires the use of OCI images. Presumably because we didn't disable that feature, buildx started to use OCI specs instead of the older Docker specs. If we disable that feature, buildx may revert back to the Docker specs? (but OCI images are the future anyway, so…)

**Comment by tsibley** [comment]
Hrumph. This also affects Singularity 3.5.3 (available on the Hutch cluster).

**Comment by tsibley** [comment]
I wondered if using an OCI registry/image URL instead of the Docker image URL would help Singularity figure it out, but alas not:

```
$ singularity build tmp.sif oras://docker.io/nextstrain/base:build-20230321T203820Z
INFO:    Starting build...
FATAL:   While performing build: conveyor failed to get: failed to get SHA of oras://docker.io/nextstrain/base:build-20230321T203820Z: could not get image manifest, received mediaType: application/vnd.oci.image.index.v1+json
```

**Comment by tsibley** [comment]
While figuring out how to best address this, I wanted to know the affected versions of Singularity. Any minimum version we might want to consider to fix this bug also interacts with minimum version considerations for a fix to #274, and the minimums there are clearer than here.

So I bisected between broken (doesn't work with OCI image indexes) and fixed (works with them) versions and found that 3.5.3 is (coincidentally!) the last broken version. 3.6.0 is the first fixed version. I then continued bisecting it down to https://github.com/sylabs/singularity/commit/d46428cc89ba22814254603f62c0c25af10f8080.

**Comment by tsibley** [comment]
Notes on bisecting:

I set up a Go build environment similar to the instructions in ... and then used the following (and minor variants to account for codebase changes) bit of shell to test each version/commit:

```bash
   ./mconfig -b builddir -p $PWD/built --without-network \
&& make -C builddir/ install \
&& ./built/bin/singularity build --disable-cache tmp.sif docker://nextstrain/base:build-20230321T203820Z
```

The final command either immediately fails or starts building the Singularity image from the Docker one (at which point I'd terminate it and mark the version/commit good).
